error/s2n_errno.c

```diff
@@ -31,209 +31,232 @@ static const char *no_such_language = "Language is not supported for error trans
-    ERROR_STRING(S2N_ERR_INVALID_DYNAMIC_THRESHOLD, "invalid dynamic record threshold")
-    ERROR_STRING(S2N_ERR_INVALID_ARGUMENT, "invalid argument provided into a function call")
-    ERROR_STRING(S2N_ERR_NOT_IN_UNIT_TEST, "Illegal configuration, can only be used during unit tests")
-    ERROR_STRING(S2N_ERR_BAD_FD, "Invalid file descriptor")
-    ERROR_STRING(S2N_ERR_RDRAND_FAILED, "Error executing rdrand instruction")
-    ERROR_STRING(S2N_ERR_UNSUPPORTED_CPU, "Unsupported CPU architecture")
-    ERROR_STRING(S2N_ERR_FAILED_CACHE_RETRIEVAL, "Failed cache retrieval")
-    ERROR_STRING(S2N_ERR_CONNECTION_CACHING_DISALLOWED, "This connection is not allowed to be cached")
-    ERROR_STRING(S2N_ERR_SESSION_TICKET_NOT_SUPPORTED, "Session ticket not supported for this connection")
-    ERROR_STRING(S2N_ERR_X509_TRUST_STORE, "Error initializing trust store")
-    ERROR_STRING(S2N_ERR_UNKNOWN_PROTOCOL_VERSION, "Error determining client protocol version")
-    ERROR_STRING(S2N_ERR_OCSP_NOT_SUPPORTED, "OCSP stapling was requested, but is not supported")
-    ERROR_STRING(S2N_ERR_NULL_CN_NAME, "Error parsing CN names")
-    ERROR_STRING(S2N_ERR_NULL_SANS, "Error parsing SANS")
-    ERROR_STRING(S2N_ERR_CLIENT_HELLO_VERSION, "Could not get client hello version")
-    ERROR_STRING(S2N_ERR_CLIENT_PROTOCOL_VERSION, "Could not get client protocol version")
-    ERROR_STRING(S2N_ERR_SERVER_PROTOCOL_VERSION, "Could not get server protocol version")
-    ERROR_STRING(S2N_ERR_ACTUAL_PROTOCOL_VERSION, "Could not get actual protocol version")
-    ERROR_STRING(S2N_ERR_POLLING_FROM_SOCKET, "Error polling from socket")
-    ERROR_STRING(S2N_ERR_RECV_STUFFER_FROM_CONN, "Error receiving stuffer from connection")
-    ERROR_STRING(S2N_ERR_SEND_STUFFER_TO_CONN, "Error sending stuffer to connection")
+#define S2N_ERROR_TYPE_ERROR_COUNT(TYPE) (TYPE##_END - TYPE##_START)
+#define ADD_ERROR(error, msg) [error & ((1 << S2N_ERR_NUM_VALUE_BITS) - 1)] = { .error_name = (#error), .error_message = (msg) }
+
+struct s2n_error_translation S2N_ERR_T_OK_EN[S2N_ERROR_TYPE_ERROR_COUNT(S2N_ERR_T_OK)] = {
+    ADD_ERROR(S2N_ERR_OK, "no error"),
+};
+
+struct s2n_error_translation S2N_ERR_T_IO_EN[S2N_ERROR_TYPE_ERROR_COUNT(S2N_ERR_T_IO)] = {
+    ADD_ERROR(S2N_ERR_IO, "underlying I/O operation failed, check system errno"),
+};
+
+struct s2n_error_translation S2N_ERR_T_CLOSED_EN[S2N_ERROR_TYPE_ERROR_COUNT(S2N_ERR_T_CLOSED)] = {
+    ADD_ERROR(S2N_ERR_CLOSED, "connection is closed"),
+};
+
+struct s2n_error_translation S2N_ERR_T_BLOCKED_EN[S2N_ERROR_TYPE_ERROR_COUNT(S2N_ERR_T_BLOCKED)] = {
+    ADD_ERROR(S2N_ERR_BLOCKED, "underlying I/O operation would block"),
+};
+
+struct s2n_error_translation S2N_ERR_T_ALERT_EN[S2N_ERROR_TYPE_ERROR_COUNT(S2N_ERR_T_ALERT)] = {
+    ADD_ERROR(S2N_ERR_ALERT, "TLS alert received"),
+};
+
+struct s2n_error_translation S2N_ERR_T_PROTO_EN[S2N_ERROR_TYPE_ERROR_COUNT(S2N_ERR_T_PROTO)] = {
+    ADD_ERROR(S2N_ERR_ENCRYPT, "error encrypting data"),
```

Comment:
what happens if the msg description is added to the wrong category, does the tests catches it?

Reply:
Yup, test will catch a gap in a category where the error should have been originally added.

Compiler catches a few things too:
1. If index is out of bound, gcc/clang will fail compilation
2. If there are duplicated indexes, clang will fail compilation